index.js

```diff
@@ -42,25 +44,28 @@ function basicPlugin (fastify, opts, next) {
   }
 }
 
-function getAuthenticateHeader (authenticate, next) {
+function getAuthenticateHeader (authenticate) {
   if (!authenticate) return false
   if (authenticate === true) {
-    return {
-      key: 'WWW-Authenticate',
-      value: 'Basic'
-    }
+    return 'Basic'
   }
   if (typeof authenticate === 'object') {
-    const realm = (authenticate.realm && typeof authenticate.realm === 'string')
-      ? authenticate.realm
-      : ''
-    return {
-      key: 'WWW-Authenticate',
-      value: 'Basic' + (realm ? ` realm="${realm}"` : '')
+    const realm = authenticate.realm
+    switch (typeof realm) {
+      case 'undefined':
+        return 'Basic'
+      case 'boolean':
+        return 'Basic'
+      case 'string':
+        return `Basic realm="${realm}"` 
+      case 'function':
+        return function (req) {
+          return `Basic realm="${realm(req)}"` 
```

Comment:
Should we support async function for `realm` ?

Reply:
I'm not convinced of the API either. I think it might be better to have `validate` provide the realm.